zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1217 @@
-pub fn init<S>(
-    state_service: S,
-    checkpoint_list: impl Into<Arc<HashMap<BlockHeight, BlockHeaderHash>>>,
-) -> impl Service<
-    Arc<Block>,
-    Response = BlockHeaderHash,
-    Error = Error,
-    Future = impl Future<Output = Result<BlockHeaderHash, Error>>,
-> + Send
-       + 'static
-where
-    S: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>
-        + Send
-        + Clone
-        + 'static,
-    S::Future: Send + 'static,
-{
-    CheckpointVerifier {
-        state_service,
-        checkpoint_list: checkpoint_list.into(),
-    }
-}
-
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::tests::install_tracing;
 
-    use color_eyre::eyre::{bail, eyre, Report};
-    use tower::{util::ServiceExt, Service};
+    use color_eyre::eyre::{eyre, Report};
+    use std::{cmp::min, mem::drop, time::Duration};
+    use tokio::time::timeout;
+    use tower::{Service, ServiceExt};
 
     use zebra_chain::serialization::ZcashDeserialize;
 
+    /// The timeout we apply to each verify future during testing.
+    ///
+    /// The checkpoint verifier uses `tokio::sync::oneshot` channels as futures.
+    /// If the verifier doesn't send a message on the channel, any tests that
+    /// await the channel future will hang.
+    ///
+    /// This value is set to a large value, to avoid spurious failures due to
+    /// high system load.
+    const VERIFY_TIMEOUT_SECONDS: u64 = 10;
+
     #[tokio::test]
     #[spandoc::spandoc]
-    async fn checkpoint_single_item_list() -> Result<(), Report> {
-        install_tracing();
+    async fn single_item_checkpoint_list() -> Result<(), Report> {
+        zebra_test::init();
 
         let block0 =
             Arc::<Block>::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES[..])?;
         let hash0: BlockHeaderHash = block0.as_ref().into();
 
         // Make a checkpoint list containing only the genesis block
-        let genesis_checkpoint_list: HashMap<BlockHeight, BlockHeaderHash> =
+        let genesis_checkpoint_list: BTreeMap<BlockHeight, BlockHeaderHash> =
             [(block0.coinbase_height().unwrap(), hash0)]
                 .iter()
                 .cloned()
                 .collect();
 
-        let mut state_service = Box::new(zebra_state::in_memory::init());
-        let mut checkpoint_verifier = super::init(state_service.clone(), genesis_checkpoint_list);
+        let mut checkpoint_verifier =
+            CheckpointVerifier::new(genesis_checkpoint_list).map_err(|e| eyre!(e))?;
+
+        assert_eq!(
+            checkpoint_verifier.previous_checkpoint_height(),
+            BeforeGenesis
+        );
+        assert_eq!(
+            checkpoint_verifier.target_checkpoint_height(),
+            WaitingForBlocks
+        );
+        assert_eq!(checkpoint_verifier.max_checkpoint_height(), BlockHeight(0));
 
         /// Make sure the verifier service is ready
         let ready_verifier_service = checkpoint_verifier
```

Comment:
TOILI: the version of spandoc we're using now _should_ be the version where I fixed the multiple `await`s restriction, so if you wanted to you can start collapsing some of these split up future chains into single expressions again.

Reply:
I think they are easier to read as they are. But if we want to do that, let's do it after we refactor parts of the tests into utility functions. Deferred to #610.